File: client/go/util/commit/commitutil.go
// Copyright 2017 LINE Corporation
//
// LINE Corporation licenses this file to you under the Apache License,
// version 2.0 (the "License"); you may not use this file except in compliance
// with the License. You may obtain a copy of the License at:
//
//   https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
// License for the specific language governing permissions and limitations
// under the License.

// Package commit implements utilities that parse the input and make commit messages.
package commit

import (
	"bufio"
	"bytes"
	"errors"
	"fmt"
	"html/template"
	"os"
	"regexp"
	"strings"
	"unicode"

	"github.com/line/centraldogma/client/go/json"
	"github.com/line/centraldogma/client/go/util/file"
)

type CommitType int

const (
	_ CommitType = iota
	Add
	Edit
	Remove
)

var tempFileName = "commit-message.txt"

func GetMessage(filePathToBeChanged string, commitType CommitType) (*json.CommitMessage, error) {
	tempFilePath, fd, err := file.NewTemp(tempFileName)
	if err != nil {
		return nil, err
	}
	defer os.Remove(tempFilePath)
	tmpl, _ := template.New("").Parse(commitMessageTemplate)
	tmpl.Execute(fd, typeWithFile{FilePath: filePathToBeChanged, CommitType: commitType})
	fd.Close()
	cmd := file.CmdToOpenEditor(tempFilePath)
	if err = cmd.Start(); err != nil {
		// Failed to launch the editor.
		return messageFromCLI()
	} else {
		err = cmd.Wait()
		if err != nil {
			return nil, errors.New("failed to write the commit message")
		}
		return messageFrom(tempFilePath)
	}
}

func messageFromCLI() (*json.CommitMessage, error) {
	reader := bufio.NewReader(os.Stdin)
	fmt.Print("Enter summary: ")
	summary, _ := reader.ReadString('\n')

	if len(summary) == 0 {
		return nil, errors.New("you must input summary")
	}
	fmt.Print("\nEnter detailed comment: ")
	content, _ := reader.ReadString('\n')
	comment := &json.Comment{Content: content, Markup: "PLAINTEXT"}
	commitMessage := &json.CommitMessage{Summary: summary, Detail: comment}
	return commitMessage, nil
}

func messageFrom(filePath string) (*json.CommitMessage, error) {
	fd, err := os.Open(filePath)
	if err != nil {
		return nil, err
	}
	defer fd.Close()
	scanner := bufio.NewScanner(fd)

	summary, err := getSummary(scanner)
	if err != nil {
		return nil, err
	}
	comment := &json.Comment{Content: getContent(scanner), Markup: "PLAINTEXT"}
	commitMessage := &json.CommitMessage{Summary: summary, Detail: comment}
	return commitMessage, nil
}

func getSummary(scanner *bufio.Scanner) (string, error) {
	line := ""
	for scanner.Scan() {
		line = strings.TrimSpace(scanner.Text())
		if !strings.HasPrefix(line, "#") && len(line) != 0 {
			return line, nil
		}
	}
	return "", errors.New("must input the summary")
}

func getContent(scanner *bufio.Scanner) string {
	var buf bytes.Buffer
	passedEmptyLinesUnderSummary := false
	for scanner.Scan() {
		line := strings.TrimRightFunc(scanner.Text(), unicode.IsSpace)
		if strings.HasPrefix(line, "#") {
			continue
		}
		if len(line) == 0 && !passedEmptyLinesUnderSummary {
			continue
		} else {
			passedEmptyLinesUnderSummary = true
		}
		buf.WriteString(line)
		buf.WriteString("\n")
	}
	// remove trailing empty lines
	regex, _ := regexp.Compile("\\n{2,}\\z")
	return regex.ReplaceAllString(buf.String(), "")
}

type typeWithFile struct {
	FilePath   string
	CommitType CommitType
}

var commitMessageTemplate = `
# Please enter the commit message for your changes. Lines starting\n
# with '#' will be ignored, and an empty message aborts the commit.\n
#
# Changes to be committed:
#   {{if eq .CommitType 1}}new file{{else if eq .CommitType 2}}modified{{else}}deleted{{end}}: {{.FilePath}}
#
`

File: server/src/main/java/com/linecorp/centraldogma/server/internal/storage/repository/git/CommitUtil.java
/*
 * Copyright 2017 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package com.linecorp.centraldogma.server.internal.storage.repository.git;

import static java.util.Objects.requireNonNull;

import java.io.IOException;
import java.io.StringWriter;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.JsonNode;

import com.linecorp.centraldogma.common.Author;
import com.linecorp.centraldogma.common.Commit;
import com.linecorp.centraldogma.common.Markup;
import com.linecorp.centraldogma.common.Revision;
import com.linecorp.centraldogma.internal.Jackson;
import com.linecorp.centraldogma.server.internal.storage.StorageException;

final class CommitUtil {

    private static final String FIELD_NAME_SUMMARY = "summary";
    private static final String FIELD_NAME_DETAIL = "detail";
    private static final String FIELD_NAME_MARKUP = "markup";
    private static final String FIELD_NAME_REVISION = "revision";

    static String toJsonString(String summary, String detail, Markup markup, Revision nextRevision) {
        try {
            StringWriter stringWriter = new StringWriter();
            JsonGenerator jsonGenerator = Jackson.createPrettyGenerator(stringWriter);
            jsonGenerator.writeStartObject();
            jsonGenerator.writeStringField(FIELD_NAME_SUMMARY, summary);
            jsonGenerator.writeStringField(FIELD_NAME_DETAIL, detail);
            jsonGenerator.writeStringField(FIELD_NAME_MARKUP, markup.nameLowercased());
            jsonGenerator.writeStringField(FIELD_NAME_REVISION, nextRevision.text());
            jsonGenerator.writeEndObject();
            jsonGenerator.close();
            return stringWriter.toString();
        } catch (IOException e) {
            throw new StorageException("failed to generate a JSON string", e);
        }
    }

    static Revision extractRevision(String jsonString) {
        try {
            JsonNode jsonNode = Jackson.readTree(jsonString);
            return new Revision(Jackson.textValue(jsonNode.get(FIELD_NAME_REVISION), ""));
        } catch (Exception e) {
            throw new StorageException("failed to extract revision from " + jsonString, e);
        }
    }

    static Commit newCommit(Author author, long when, String jsonString) {
        requireNonNull(author, "author");
        when = when / 1000L * 1000L; // Drop the milliseconds
        try {
            JsonNode jsonNode = Jackson.readTree(jsonString);

            final String summary = Jackson.textValue(jsonNode.get(FIELD_NAME_SUMMARY), "");
            final String detail = Jackson.textValue(jsonNode.get(FIELD_NAME_DETAIL), "");

            final Markup markup;
            switch (Jackson.textValue(jsonNode.get(FIELD_NAME_MARKUP), "")) {
            case "plaintext":
                markup = Markup.PLAINTEXT;
                break;
            case "markdown":
                markup = Markup.MARKDOWN;
                break;
            default:
                markup = Markup.UNKNOWN;
            }

            final Revision revision = new Revision(Jackson.textValue(jsonNode.get(FIELD_NAME_REVISION), ""));

            return new Commit(revision, author, when, summary, detail, markup);
        } catch (Exception e) {
            throw new StorageException("failed to create a Commit", e);
        }
    }

    private CommitUtil() {}
}
